### PR TITLE
Cleanup the symantics of ACS Handler

### PR DESCRIPTION
**Pull request by aws-gibbskt**
### Summary
The ACS handler was using a channel to support looping, but the
channel was only sent to from the current thread via a go routine
that would just write to the channel.

The rewrite maintains the behavior of always connecting when
disconnected potentially after a retry backoff. However it uses
a simple endless for loop rather than reading from channels to
mimic this straightforward behavior.

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: agent/acs/handler/acs_handler.go [hunk 1]
```diff
@@ -189,70 +189,59 @@ func NewSession(
...
+				// If the context was not cancelled and we've waited for the
+				// wait duration without any errors, send the message to the channel
+				// to reconnect to ACS
+				seelog.Info("Done waiting; reconnecting to ACS")
+			} else {
+				// Wait was interrupted. We expect the session to close as canceling
+				// the session context is the only way to end up here. Print a message
+				// to indicate the same
+				seelog.Info("Interrupted waiting for reconnect delay to elapse; Expect session to close")
+				return nil
```
Comment (nodir-t) [inline agent/acs/handler/acs_handler.go:242]: Note that this isn't equivalent to the prev revision. This exits without waiting for the context to get canceled. The equivalent version would be do `<- ctx.Done()` before returning

File: agent/acs/handler/acs_handler.go [hunk 2]
```diff
@@ -189,70 +189,59 @@ func NewSession(
...
+			// If ACS closed the connection, there's no need to backoff,
+			// reconnect immediately
+			seelog.Infof("ACS Websocket connection closed for a valid reason: %v", acsError)
+			acsSession.backoff.Reset()
+		} else {
+			// Disconnected unexpectedly from ACS, compute backoff duration to
+			// reconnect
+			reconnectDelay := acsSession.computeReconnectDelay(isInactiveInstance)
+			seelog.Infof("Reconnecting to ACS in: %s", reconnectDelay.String())
+			waitComplete := acsSession.waitForDuration(reconnectDelay)
```
Comment (nodir-t) [inline agent/acs/handler/acs_handler.go:231]: nit: inline this variable?

**Review by nodir-t — CHANGES_REQUESTED** [review]

File: agent/acs/handler/acs_handler.go (hunk 1 above)
Comment (aws-gibbskt) [reply agent/acs/handler/acs_handler.go:242]: It actually is. In the sense that Done() being signaled is the only way that false is returned. We could check it again here if you want but it isn't any different.

Comment (nodir-t) [reply agent/acs/handler/acs_handler.go:242]: you are right

File: agent/acs/handler/acs_handler.go [hunk 3]
```diff
@@ -189,70 +189,59 @@ func NewSession(
 // Start starts the session. It'll forever keep trying to connect to ACS unless
 // the context is cancelled.
 //
-// If the context is cancelled, Start() would return with the error code returned
-// by the context.
+// If the context is cancelled, Start() would return nil
```
Comment (nodir-t) [inline agent/acs/handler/acs_handler.go:192]: seems like it returns nil unconditionally. Might as well remove the return type.

**Review by nodir-t — APPROVED** [review]

**Comment by angelcar** [comment]
I don't think this code is equivalent? If I am not mistaken, this loop will execute over and over since nothing is pausing the loop execution -- the `default` clause causes the `case <-acsSession.ctx.Done()` to be ignored.

**Comment by aws-gibbskt** [comment]
Oops yeah I closed this one in favor of https://github.com/aws/amazon-ecs-agent/pull/3225 which is targeting dev not a feature branch... @nodir-t @angelcar

**Pull request closed without merging**